### ISSUE TITLE
Project specific cd_user

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-provisioning-app#98]
We had some issues with a global `cd_user` (e.g. if password is changed, there are lots of places that need to be adjusted). This would be easier if each project had its own `cd_user`. 

Just raising this question / issue here - not sure how that would work and what implications this has.

FYI @clemensutschig @gerardcl @stitakis @rattermeyer

**Comment by clemensutschig** [comment]
@michaelsauter- i have this some thought over the weekend.. and the blocker is rundeck. Crazy idea - What of we move the whole QS magic into jenkins which already lives in openshift. Like seed a Master creation Pipeline into a Project-cd namespace- that is triggered via Webhook Proxy and can do all the rundeck qs capability

**Comment by gerardcl** [comment]
I like it! would this be based on temporary pipelines? If yes, then we must keep those pipelines in a write-once mode (so to keep traceability) and not deleteable. Does it make sense? I'm just keeping the thought on being based on a master pipeline in CD project...

**Comment by clemensutschig** [comment]
@gerardcl - i would keep static ones (in fact only one) - as simple as possible .. that could refer to a Master jenkins file in ods-project quickstarters

**Comment by michaelsauter** [comment]
@clemensutschig Sounds interesting! Definitely something to think about. Maybe we could kill Rundeck completely? I'm not a huge fan and its role overlaps with the Provisioning App and, if we make your proposed change, Jenkins pipelines.

I will think a bit about this during the week ...

**Comment by stitakis** [comment]
@clemensutschig I´m struggling to follow here... what is QS and the QS magic?

**Comment by clemensutschig** [comment]
@stitakis  QS => QuickStarters

**Comment by stitakis** [comment]
@clemensutschig ah yes... idea sounds good to me. We will be able to reduce moving parts and complexity, right?

**Comment by clemensutschig** [comment]
@stitakis - we will kick out rundeck :)  (@maksymmaksym  fyi)

**Comment by gerardcl** [comment]
is this going into 2.0.x?

**Comment by michaelsauter** [comment]
One thing that came to mind: This requires the *-cd project to have write permissions to the *-dev and *-test environments ... I know we already have this (ad-hoc cloning), but doing the proposed change makes this permission really really necessary, and undermines the idea of confining attacks within one namespace. Just putting it out there as a negative consequence ...

**Comment by gerardcl** [comment]
Thinking loud: what about an approval step? (Like PR/notification) from a human...

**Comment by clemensutschig** [comment]
@michaelsauter - how would that work in reality -  you'd have to gain admin rights on cd to travel into test / dev - that would only work if you are able to get the secret in the slave build pod

**Comment by tjaeschke** [comment]
So just a few thoughts before we start implementing this.

As we said, user creation will be out of scope, because it depends on business processes, that are not handled by ODS.

The flow:
1. create cd user for project
2. Pass cd user name in prov app (alternatively use a naming pattern?)
3. Add user to Bitbucket project
4. Set cd user secret in project and for project jenkins

Parts that have to be updated:
* Project creation in OpenShift
* Bitbucket project creation
* Provisioning App UI?

No password handling after all, Alternatives like personal access token will be evaluated later.
